### PR TITLE
Bohdan

### PR DESCRIPTION
**Pull request by BogdanKalysh**
Add transaction functionality in frontend

### CODE REVIEW

**Review by Pasemko — APPROVED** [review]

File: budget_manager_frontend/headers/loginWindow.h [hunk 1]
```diff
@@ -30,7 +30,7 @@ private slots:
     Ui::LoginWindow *ui;
     MainWindow *mainWindow;
     SignUpWindow *signUpWindow;
-
+    QNetworkAccessManager* manager = new QNetworkAccessManager;
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/headers/loginWindow.h:33]: It is better to use smart pointers

File: budget_manager_frontend/headers/parsers/transactionparser.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#ifndef TRANSACTIONPARSER_H
+#define TRANSACTIONPARSER_H
+
+#include "ijsonparser.h"
+#include "transaction.h"
+
+
+class TransactionParser: IJsonParser<Transaction>
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/headers/parsers/transactionparser.h:8]: You do not inherit the interface. You inherit implementation only.

File: budget_manager_frontend/headers/parsers/userparser.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#ifndef USERPARSER_H
+#define USERPARSER_H
+
+#include "ijsonparser.h"
+#include "user.h"
+
+
+class UserParser: IJsonParser<User>
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/headers/parsers/userparser.h:8]: You do not inherit the interface. You inherit implementation only.

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 4]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+        QJsonDocument jsonResponse = QJsonDocument::fromJson(categoriesStr.toUtf8());
+        QJsonArray json_array = jsonResponse.array();
+
+        CategoryParser parser;
+        categories = parser.parseVector(json_array);
+    }
+
+    ui->cetegoryComboBox->clear();
+
+    if(ui->expenceRadioButton->isChecked()){
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:53]: Code is duplicating and very weird logic. Please simplify this code.

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 5]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+void MainWindow::readCategories()
+{
+    QNetworkReply *categoriesReply = qobject_cast<QNetworkReply*>(sender());
+
+    if(categoriesReply){
+        QString categoriesStr = categoriesReply->readAll();
+        QJsonDocument jsonResponse = QJsonDocument::fromJson(categoriesStr.toUtf8());
+        QJsonArray json_array = jsonResponse.array();
+
+        CategoryParser parser;
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:47]: You are inheriting from IJsonParser. You should use this interface.

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 6]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+void MainWindow::readTransactions()
+{
+    QNetworkReply *transactionsReply = qobject_cast<QNetworkReply*>(sender());
+
+    if(transactionsReply){
+        QString transactionsStr = transactionsReply->readAll();
+        QJsonDocument jsonResponse = QJsonDocument::fromJson(transactionsStr.toUtf8());
+        QJsonArray json_array = jsonResponse.array();
+
+        TransactionParser parser;
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:79]: You should use polymorphism.

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 7]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+    postTranasactionReply->close();
+    postTranasactionReply->deleteLater();
+}
+
 
 void MainWindow::on_addTransactionButton_clicked()
 {
-    int amount = ui->amountInputLine->text().toInt();
+
+    int amount;
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:107]: Please assign a default value.

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 8]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+    if (ui->incomeRadioButton->isChecked())
+        amount = ui->amountInputLine->text().toInt();
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:109]: I suppose there should be +

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 9]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+    else
+        amount = -ui->amountInputLine->text().toInt();
+
     QString description = ui->descriptionInputLine->text();
     QString category = ui->cetegoryComboBox->currentText();
+    int categoryId(0);
 
-    qDebug() << amount << " " << description << " " << category << "\n";
-}
+    foreach(Category cat, categories){
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:117]: Please use https://en.cppreference.com/w/cpp/language/range-for

Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:117]: generally better to use algorithms for instance find_if

File: budget_manager_frontend/sources/models/transaction.cpp [hunk 10]
```diff
@@ -1,15 +1,28 @@
...
 
-Transaction::Transaction(int id, int amount, QDate date, QString description, int categoryId)
+
+Transaction::Transaction(int id, int amount, QString description, int categoryId)
+{
+    this->id = id;
+    this->amount = amount;
+    this->description = description;
+    this->categoryId = categoryId;
+
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/models/transaction.cpp:11]: remove empty line

File: budget_manager_frontend/sources/parsers/transactionparser.cpp [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    int id = json.value(jsonbuilder::ID).toInt();
+    int amount = json.value(jsonbuilder::AMOUNT).toInt();
+    QDate date = json.value(jsonbuilder::DATE).toVariant().toDate();
+    QString description = json.value(jsonbuilder::DESCRIPTION).toString();
+
+    int categoryId = json.value(jsonbuilder::CATEGORYID).toInt();
+    QString categoryName = json.value(jsonbuilder::CATEGORYNAME).toString();
+    QColor color(json.value(jsonbuilder::COLOR).toString());
+
+
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/parsers/transactionparser.cpp:15]: remove empty line

File: budget_manager_frontend/sources/parsers/transactionparser.cpp [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    return Transaction(id, amount, date, description, categoryId, categoryName, color);
+}
+
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/parsers/transactionparser.cpp:18]: remove empty lines

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 13]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+
+void MainWindow::finishedPostTransactions()
+{
+    QNetworkReply *postTranasactionReply = qobject_cast<QNetworkReply*>(sender());
+
+    if(postTranasactionReply->error() == QNetworkReply::NoError){
+        QString contents = QString::fromUtf8(postTranasactionReply->readAll());
+        qDebug() << contents;
+    }
+    else{
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:95]: is this format for curly braces correct?

File: budget_manager_frontend/sources/mainWindow.cpp [hunk 14]
```diff
@@ -20,33 +35,121 @@ MainWindow::~MainWindow()
...
+    }
+
+    ui->cetegoryComboBox->clear();
+
+    if(ui->expenceRadioButton->isChecked()){
+        foreach(Category cat, categories){
+            if(!cat.getType())
+                ui->cetegoryComboBox->addItem(cat.getName());
+        }
+    }else{
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/sources/mainWindow.cpp:58]: is this format for curly braces correct?

File: budget_manager_frontend/headers/mainWindow.h [hunk 15]
```diff
@@ -12,18 +17,28 @@ class MainWindow : public QMainWindow
...
 public:
-    MainWindow(QWidget *parent = nullptr);
+    MainWindow(User user, QNetworkAccessManager* manager,  QWidget *parent=nullptr);
     ~MainWindow();
 
+public slots:
+    void readCategories();
+    void readTransactions();
+    void finishedPostTransactions();
+
```
Comment (AndriyFedorenko) [inline budget_manager_frontend/headers/mainWindow.h:27]: remove empty line

**Review by AndriyFedorenko — CHANGES_REQUESTED** [review]
Please read about polymorphism in C++ and how to use Interfaces. 
Pay attention to code formatting and some redundant empty lines and spaces

File: budget_manager_frontend/headers/builders/transactionJsonBuilder.h [hunk 16]
```diff
@@ -1,4 +1,15 @@
 #ifndef TRANSACTIONJSONBUILDER_H
 #define TRANSACTIONJSONBUILDER_H
 
+#include "IJsonBuilder.h"
+#include <QJsonObject>
+#include <QJsonValue>
+#include "transaction.h"
+
+class TransactionJsonBuilder: IJsonBuilder<Transaction>
```
[the remark this answers is not in the document]
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/builders/transactionJsonBuilder.h]: Changed inherit mode to public

File: budget_manager_frontend/headers/builders/userJsonBuilder.h [hunk 17]
```diff
@@ -1,15 +1,16 @@
...
+#ifndef USERJSONBUILDER_H
+#define USERJSONBUILDER_H
 #include "IJsonBuilder.h"
 #include <QJsonObject>
 #include <QJsonValue>
 #include "user.h"
 
 
-class LoginJsonBuilder: IJsonBuilder<User>
+class UserJsonBuilder: IJsonBuilder<User>
```
[the remark this answers is not in the document]
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/builders/userJsonBuilder.h]: Changed inherit mode to public

File: budget_manager_frontend/headers/loginWindow.h (hunk 1 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/loginWindow.h]: Used QSharedPointer for this

File: budget_manager_frontend/headers/mainWindow.h (hunk 15 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/mainWindow.h:29]: removed

File: budget_manager_frontend/headers/parsers/transactionparser.h (hunk 2 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/parsers/transactionparser.h]: Changed inherit mode to public

File: budget_manager_frontend/headers/parsers/userparser.h (hunk 3 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/headers/parsers/userparser.h]: Changed inherit mode to public

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 5 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: Used QSharedPointer for parser and made it IJsonParser initialised as CategoryParser

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 4 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: Removed code duplication and simplified

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 14 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: fixed braces format

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 6 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: Used IJsonParser

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 13 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: fixed braces format

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 7 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: fixed

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 8 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: Removed if statement at all, it should be like this

File: budget_manager_frontend/sources/mainWindow.cpp (hunk 9 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: used for instead of foreach everywhere

Comment (BogdanKalysh) [reply budget_manager_frontend/sources/mainWindow.cpp]: Found way to just get id from combobox without searching it in vector

File: budget_manager_frontend/sources/models/transaction.cpp (hunk 10 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/models/transaction.cpp]: Removed

File: budget_manager_frontend/sources/parsers/transactionparser.cpp (hunk 11 above)
Comment (BogdanKalysh) [reply budget_manager_frontend/sources/parsers/transactionparser.cpp]: Removed

**Review by Nazar-TK — COMMENTED** [review]
Good

**Pull request merged**